source/opt/scalar_analysis.cpp

```diff
@@ -0,0 +1,386 @@
+
+  return CreateMultiplyNode(op1, op2);
+}
+
+SENode* ScalarEvolutionAnalysis::CreateMultiplyNode(SENode* operand_1,
+                                                    SENode* operand_2) {
+  if (operand_1->GetType() == SENode::Constant &&
+      operand_2->GetType() == SENode::Constant) {
+    return CreateConstant(operand_1->AsSEConstantNode()->FoldToSingleValue() *
+                          operand_2->AsSEConstantNode()->FoldToSingleValue());
+  }
+
+  std::unique_ptr<SENode> add_node{new SEMultiplyNode(this)};
+
+  add_node->AddChild(operand_1);
+  add_node->AddChild(operand_2);
+
+  return GetCachedOrAdd(std::move(add_node));
+}
+
+SENode* ScalarEvolutionAnalysis::CreateSubtraction(SENode* operand_1,
+                                                   SENode* operand_2) {
+  return CreateAddNode(operand_1, CreateNegation(operand_2));
+}
+
+SENode* ScalarEvolutionAnalysis::CreateAddNode(SENode* operand_1,
+                                               SENode* operand_2) {
+  std::unique_ptr<SENode> add_node{new SEAddNode(this)};
+
+  add_node->AddChild(operand_1);
+  add_node->AddChild(operand_2);
+
+  return GetCachedOrAdd(std::move(add_node));
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzeInstruction(
+    const ir::Instruction* inst) {
+  if (instruction_map_.find(inst) != instruction_map_.end())
+    return instruction_map_[inst];
+
+  SENode* output = nullptr;
+  switch (inst->opcode()) {
+    case SpvOp::SpvOpPhi: {
+      output = AnalyzePhiInstruction(inst);
+      break;
+    }
+    case SpvOp::SpvOpConstant: {
+      output = AnalyzeConstant(inst);
+      break;
+    }
+    case SpvOp::SpvOpIAdd: {
+      output = AnalyzeAddOp(inst, false);
+      break;
+    }
+    case SpvOp::SpvOpISub: {
+      output = AnalyzeAddOp(inst, true);
+      break;
+    }
+    case SpvOp::SpvOpIMul: {
+      output = AnalyzeMultiplyOp(inst);
+      break;
+    }
+    default: {
+      output = CreateValueUnknownNode(inst);
+      instruction_map_[inst] = output;
+      break;
+    }
+  }
+  return output;
+}
+
+SENode* ScalarEvolutionAnalysis::AnalyzeConstant(const ir::Instruction* inst) {
+  if (inst->NumInOperands() != 1) {
+    assert(false);
+  }
+
+  int64_t value = 0;
+
+  // Look up the instruction in the constant manager.
+  const opt::analysis::Constant* constant =
+      context_->get_constant_mgr()->FindDeclaredConstant(inst->result_id());
+
+  if (constant->AsIntConstant()->type()->AsInteger()->IsSigned()) {
+    value = constant->AsIntConstant()->GetS32BitValue();
+  } else {
+    value = constant->AsIntConstant()->GetU32BitValue();
+  }
+
+  return CreateConstant(value);
+}
+
+// Handles both addition and subtraction. If the |sub| flag is set then the
+// addition will be op1+(-op2) otherwise op1+op2.
+SENode* ScalarEvolutionAnalysis::AnalyzeAddOp(const ir::Instruction* add,
+                                              bool sub) {
+  opt::analysis::DefUseManager* def_use = context_->get_def_use_mgr();
```

Comment:
Assert |add| is a supported add instruction.

Reply:
The flag |sub| seems unnecessary, the instruction's opcode can be used to decide.